core/src/main/java/org/renjin/primitives/S3.java

```diff
@@ -258,6 +268,314 @@ public static SEXP tryDispatchFromPrimitive(Context context, Environment rho, Fu
+        String signature = argSignature.getArgument(argumentClassIdx);
+        if(possibleSignatures.isEmpty() || possibleSignatures.toArray().length < row + 1 || possibleSignatures.toArray()[row] == null) {
+          int[] distance = argSignature.getDistanceAsArray(argumentClassIdx);
+          possibleSignatures.add(row, new MethodRanking(signature, distance));
+        } else {
+          int distance = argSignature.getDistance(argumentClassIdx);
+          possibleSignatures.set(row, possibleSignatures.get(row).append(signature, distance));
+        }
+        if(repeat == 1) {
+          argumentClassIdx++;
+        }
+        if(repeat != 1 && repeatIdx == repeat) {
+          repeatIdx = 0;
+          argumentClassIdx++;
+        }
+      }
+      repeatIdx = 1;
+      argumentClassIdx = 0;
+      repeat = repeat * numberOfClassesCurrentArgument;
+    }
+    return possibleSignatures;
+  }
+  
+  private static String evaluateAndGetClass(Context context, PairList args, Environment rho, int argumentIndex) {
+    // To get the class of an argument we have to evaluate the argument first. If its an atomic, than it lacks class
+    // attribute. In this case we use Attributes.getClass() to get the class name.
+    String className;
+    SEXP evaluatedArg = context.evaluate(args.getElementAsSEXP(argumentIndex), rho);
+    if(evaluatedArg.getAttributes().hasClass()) {
+      className = evaluatedArg.getAttributes().getClassVector().getElementAsString(0);
+    } else {
+      className = Attributes.getClass(evaluatedArg).getElementAsString(0);
+    }
+    return className;
+  }
+  
+  
+  /**
+   * Each class is stored as an S4 object with prefix '.__C__'. This S4 object
+   * has a attribute 'contains' where the superclass names and distance are stored.
+   * The first class is the input class with distance of 0. All of the super classes
+   * and their distance are added using the names of 'contains' and the accompanying
+   * value in 'distance' slot. "ANY" is added as the last class and the distance is
+   * defined as the maximum distance + 1. The class names and distances are stored in
+   * an ArgumentSignature class and return.
+   *
+   * */
+
+  private static ArgumentSignature getAllClassAndDistance(Context context, String argClass) {
+    Symbol argClassObjectName = Symbol.get(".__C__" + argClass);
+    Frame globalFrame = context.getGlobalEnvironment().getFrame();
+    AttributeMap map = globalFrame.getVariable(argClassObjectName).getAttributes();
+    SEXP containsSlot = map.get("contains");
+    SEXP argSuperClasses = containsSlot.getNames();
+    int[] distances = new int[argSuperClasses.length() + 2];
+    String[] classes = new String[argSuperClasses.length() + 2];
+
+    classes[0] = argClass;
+    distances[0] = 0;
+    for(int i = 0; i < argSuperClasses.length(); i++) {
+      DoubleArrayVector distanceSlot = (DoubleArrayVector) ((ListVector) containsSlot).get(i).getAttributes().get("distance");
```

Comment:
Why are you assuming this is a `DoubleArrayVector` ? This may throw a `ClassCastException` at runtime.